File: ts/tests/actor.test.ts
import { describe, it, expect } from "vitest";
import {
  Actor,
  ActorType,
  ActorStatus,
  InMemoryActorStore,
  transitionTo,
  validTransitions,
} from "../src/actor.js";
import { ActorId, EventId, PublicKey } from "../src/types.js";
import {
  ActorNotFoundError,
  ActorKeyNotFoundError,
  InvalidTransitionError,
} from "../src/errors.js";

function testPublicKey(b: number): PublicKey {
  const key = new Uint8Array(32);
  key[0] = b;
  return new PublicKey(key);
}

const reason = new EventId("019462a0-0000-7000-8000-000000000001");

describe("InMemoryActorStore", () => {
  it("Register", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    expect(a.displayName).toBe("Alice");
    expect(a.actorType).toBe(ActorType.Human);
    expect(a.status).toBe(ActorStatus.Active);
  });

  it("RegisterIdempotent", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a1 = s.register(pk, "Alice", ActorType.Human);
    const a2 = s.register(pk, "Alice Again", ActorType.Human);
    expect(a1.id.value).toBe(a2.id.value);
    expect(s.actorCount()).toBe(1);
  });

  it("Get", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    const got = s.get(a.id);
    expect(got.id.value).toBe(a.id.value);
  });

  it("GetNotFound", () => {
    const s = new InMemoryActorStore();
    expect(() => s.get(new ActorId("actor_nonexistent"))).toThrow(ActorNotFoundError);
  });

  it("GetByPublicKey", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    const got = s.getByPublicKey(pk);
    expect(got.id.value).toBe(a.id.value);
  });

  it("GetByPublicKeyNotFound", () => {
    const s = new InMemoryActorStore();
    expect(() => s.getByPublicKey(testPublicKey(99))).toThrow(ActorKeyNotFoundError);
  });

  it("Update", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    const updated = s.update(a.id, { displayName: "Alice Updated" });
    expect(updated.displayName).toBe("Alice Updated");
  });

  it("UpdateMetadataMerge", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);

    // Set initial metadata
    s.update(a.id, { metadata: { role: "builder" } });

    // Merge additional metadata
    const updated = s.update(a.id, { metadata: { team: "core" } });
    const md = updated.metadata;
    expect(md["role"]).toBe("builder");
    expect(md["team"]).toBe("core");
  });

  it("UpdateNotFound", () => {
    const s = new InMemoryActorStore();
    expect(() => s.update(new ActorId("actor_nonexistent"), {})).toThrow(ActorNotFoundError);
  });

  it("Suspend", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    const suspended = s.suspend(a.id, reason);
    expect(suspended.status).toBe(ActorStatus.Suspended);
  });

  it("SuspendNotFound", () => {
    const s = new InMemoryActorStore();
    expect(() => s.suspend(new ActorId("actor_nonexistent"), reason)).toThrow(ActorNotFoundError);
  });

  it("SuspendAndReactivate", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    s.suspend(a.id, reason);

    const got = s.get(a.id);
    expect(got.status).toBe(ActorStatus.Suspended);

    const reactivated = s.reactivate(a.id, reason);
    expect(reactivated.status).toBe(ActorStatus.Active);
  });

  it("Memorial", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    const memorial = s.memorial(a.id, reason);
    expect(memorial.status).toBe(ActorStatus.Memorial);
  });

  it("MemorialIsTerminal", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    s.memorial(a.id, reason);

    // Try to suspend — should fail
    expect(() => s.suspend(a.id, reason)).toThrow(InvalidTransitionError);
  });

  it("MemorialReactivateIsError", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    s.memorial(a.id, reason);

    expect(() => s.reactivate(a.id, reason)).toThrow(InvalidTransitionError);
  });

  it("ReactivateFromActiveIsError", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(1);
    const a = s.register(pk, "Alice", ActorType.Human);
    expect(() => s.reactivate(a.id, reason)).toThrow(InvalidTransitionError);
  });

  it("ReactivateNotFound", () => {
    const s = new InMemoryActorStore();
    expect(() => s.reactivate(new ActorId("actor_nonexistent"), reason)).toThrow(ActorNotFoundError);
  });

  it("List", () => {
    const s = new InMemoryActorStore();
    for (let i = 1; i <= 5; i++) {
      s.register(testPublicKey(i), "Actor", ActorType.Human);
    }
    const page = s.list({ limit: 10 });
    expect(page.items.length).toBe(5);
  });

  it("ListWithStatusFilter", () => {
    const s = new InMemoryActorStore();
    s.register(testPublicKey(1), "Active1", ActorType.Human);
    s.register(testPublicKey(2), "Active2", ActorType.Human);
    const a3 = s.register(testPublicKey(3), "ToBeSuspended", ActorType.Human);

    s.suspend(a3.id, reason);

    const activePage = s.list({ status: ActorStatus.Active, limit: 10 });
    expect(activePage.items.length).toBe(2);
  });

  it("ListWithTypeFilter", () => {
    const s = new InMemoryActorStore();
    s.register(testPublicKey(1), "Human", ActorType.Human);
    s.register(testPublicKey(2), "AI", ActorType.AI);
    s.register(testPublicKey(3), "System", ActorType.System);

    const page = s.list({ actorType: ActorType.AI, limit: 10 });
    expect(page.items.length).toBe(1);
  });

  it("ListPagination", () => {
    const s = new InMemoryActorStore();
    for (let i = 1; i <= 5; i++) {
      s.register(testPublicKey(i), "Actor", ActorType.Human);
    }

    // Page 1
    const page1 = s.list({ limit: 2 });
    expect(page1.items.length).toBe(2);
    expect(page1.hasMore).toBe(true);
    expect(page1.cursor).toBeDefined();

    // Page 2
    const page2 = s.list({ limit: 2, after: page1.cursor });
    expect(page2.items.length).toBe(2);
    expect(page2.hasMore).toBe(true);

    // Page 3
    const page3 = s.list({ limit: 2, after: page2.cursor });
    expect(page3.items.length).toBe(1);
    expect(page3.hasMore).toBe(false);
  });

  it("ActorGetters", () => {
    const s = new InMemoryActorStore();
    const pk = testPublicKey(42);
    const a = s.register(pk, "Alice", ActorType.AI);

    expect(a.id.value).toBeTruthy();
    expect(a.publicKey).toBeDefined();
    expect(a.createdAt).toBeGreaterThan(0);
    expect(a.metadata).toBeDefined();
    expect(typeof a.metadata).toBe("object");
  });
});

describe("ActorStatus transitions", () => {
  it("active to suspended", () => {
    expect(transitionTo(ActorStatus.Active, ActorStatus.Suspended)).toBe(ActorStatus.Suspended);
  });

  it("active to memorial", () => {
    expect(transitionTo(ActorStatus.Active, ActorStatus.Memorial)).toBe(ActorStatus.Memorial);
  });

  it("suspended to active", () => {
    expect(transitionTo(ActorStatus.Suspended, ActorStatus.Active)).toBe(ActorStatus.Active);
  });

  it("suspended to memorial", () => {
    expect(transitionTo(ActorStatus.Suspended, ActorStatus.Memorial)).toBe(ActorStatus.Memorial);
  });

  it("memorial to anything throws", () => {
    expect(() => transitionTo(ActorStatus.Memorial, ActorStatus.Active)).toThrow(InvalidTransitionError);
    expect(() => transitionTo(ActorStatus.Memorial, ActorStatus.Suspended)).toThrow(InvalidTransitionError);
  });

  it("active to active throws", () => {
    expect(() => transitionTo(ActorStatus.Active, ActorStatus.Active)).toThrow(InvalidTransitionError);
  });

  it("validTransitions returns correct targets", () => {
    expect(validTransitions(ActorStatus.Active)).toEqual([ActorStatus.Suspended, ActorStatus.Memorial]);
    expect(validTransitions(ActorStatus.Suspended)).toEqual([ActorStatus.Active, ActorStatus.Memorial]);
    expect(validTransitions(ActorStatus.Memorial)).toEqual([]);
  });
});

describe("Actor immutability", () => {
  it("metadata is deep copied on construction and access", () => {
    const md = { key: "value", nested: { a: 1 } };
    const actor = new Actor(
      new ActorId("actor_test"),
      testPublicKey(1),
      "Test",
      ActorType.Human,
      md,
      Date.now() * 1_000_000,
      ActorStatus.Active,
    );

    // Mutating original should not affect actor
    md.key = "changed";
    expect(actor.metadata["key"]).toBe("value");

    // Mutating returned metadata should not affect actor
    const got = actor.metadata;
    got["key"] = "hacked";
    expect(actor.metadata["key"]).toBe("value");
  });
});

File: rust/tests/actor_test.rs
use eventgraph::actor::*;
use eventgraph::errors::EventGraphError;
use eventgraph::types::PublicKey;

fn test_public_key(b: u8) -> PublicKey {
    let mut key = [0u8; 32];
    key[0] = b;
    PublicKey::new(key)
}

// ── Registration ──────────────────────────────────────────────────────

#[test]
fn test_register() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);

    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();
    assert_eq!(actor.display_name(), "Alice");
    assert_eq!(actor.actor_type(), ActorType::Human);
    assert_eq!(actor.status(), ActorStatus::Active);
}

#[test]
fn test_register_idempotent() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);

    let a1 = store.register(pk.clone(), "Alice", ActorType::Human).unwrap();
    let a2 = store.register(pk, "Alice Again", ActorType::Human).unwrap();

    assert_eq!(a1.id(), a2.id());
    assert_eq!(store.actor_count(), 1);
}

// ── Get ───────────────────────────────────────────────────────────────

#[test]
fn test_get() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    let got = store.get(actor.id()).unwrap();
    assert_eq!(got.id(), actor.id());
}

#[test]
fn test_get_not_found() {
    let store = InMemoryActorStore::new();
    let id = eventgraph::types::ActorId::new("actor_nonexistent").unwrap();
    let err = store.get(&id).unwrap_err();
    match err {
        EventGraphError::ActorNotFound { .. } => {}
        other => panic!("expected ActorNotFound, got {:?}", other),
    }
}

// ── GetByPublicKey ────────────────────────────────────────────────────

#[test]
fn test_get_by_public_key() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk.clone(), "Alice", ActorType::Human).unwrap();

    let got = store.get_by_public_key(&pk).unwrap();
    assert_eq!(got.id(), actor.id());
}

#[test]
fn test_get_by_public_key_not_found() {
    let store = InMemoryActorStore::new();
    let pk = test_public_key(99);
    let err = store.get_by_public_key(&pk).unwrap_err();
    match err {
        EventGraphError::ActorKeyNotFound { .. } => {}
        other => panic!("expected ActorKeyNotFound, got {:?}", other),
    }
}

// ── Update ────────────────────────────────────────────────────────────

#[test]
fn test_update() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    let updated = store.update(actor.id(), &ActorUpdate {
        display_name: Some("Alice Updated".to_string()),
        metadata: None,
    }).unwrap();
    assert_eq!(updated.display_name(), "Alice Updated");
}

#[test]
fn test_update_metadata_merge() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    // Set initial metadata
    let mut md1 = std::collections::BTreeMap::new();
    md1.insert("role".to_string(), serde_json::Value::String("builder".to_string()));
    store.update(actor.id(), &ActorUpdate {
        display_name: None,
        metadata: Some(md1),
    }).unwrap();

    // Merge additional metadata
    let mut md2 = std::collections::BTreeMap::new();
    md2.insert("team".to_string(), serde_json::Value::String("core".to_string()));
    let updated = store.update(actor.id(), &ActorUpdate {
        display_name: None,
        metadata: Some(md2),
    }).unwrap();

    let md = updated.metadata();
    assert_eq!(md.get("role").unwrap(), &serde_json::Value::String("builder".to_string()));
    assert_eq!(md.get("team").unwrap(), &serde_json::Value::String("core".to_string()));
}

#[test]
fn test_update_not_found() {
    let mut store = InMemoryActorStore::new();
    let id = eventgraph::types::ActorId::new("actor_nonexistent").unwrap();
    let err = store.update(&id, &ActorUpdate {
        display_name: None,
        metadata: None,
    }).unwrap_err();
    match err {
        EventGraphError::ActorNotFound { .. } => {}
        other => panic!("expected ActorNotFound, got {:?}", other),
    }
}

// ── Suspend ───────────────────────────────────────────────────────────

#[test]
fn test_suspend() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    let suspended = store.suspend(actor.id(), "reason-event-id").unwrap();
    assert_eq!(suspended.status(), ActorStatus::Suspended);
}

#[test]
fn test_suspend_not_found() {
    let mut store = InMemoryActorStore::new();
    let id = eventgraph::types::ActorId::new("actor_nonexistent").unwrap();
    let err = store.suspend(&id, "reason").unwrap_err();
    match err {
        EventGraphError::ActorNotFound { .. } => {}
        other => panic!("expected ActorNotFound, got {:?}", other),
    }
}

#[test]
fn test_suspend_and_reactivate() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    store.suspend(actor.id(), "reason").unwrap();
    let got = store.get(actor.id()).unwrap();
    assert_eq!(got.status(), ActorStatus::Suspended);

    let reactivated = store.reactivate(actor.id(), "reason").unwrap();
    assert_eq!(reactivated.status(), ActorStatus::Active);
}

// ── Memorial ──────────────────────────────────────────────────────────

#[test]
fn test_memorial() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    let memorial = store.memorial(actor.id(), "reason").unwrap();
    assert_eq!(memorial.status(), ActorStatus::Memorial);
}

#[test]
fn test_memorial_is_terminal() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    store.memorial(actor.id(), "reason").unwrap();

    // Try to suspend — should fail
    let err = store.suspend(actor.id(), "reason").unwrap_err();
    match err {
        EventGraphError::InvalidTransition { .. } => {}
        other => panic!("expected InvalidTransition, got {:?}", other),
    }
}

#[test]
fn test_memorial_reactivate_is_error() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    store.memorial(actor.id(), "reason").unwrap();

    let err = store.reactivate(actor.id(), "reason").unwrap_err();
    match err {
        EventGraphError::InvalidTransition { .. } => {}
        other => panic!("expected InvalidTransition, got {:?}", other),
    }
}

// ── Reactivate ────────────────────────────────────────────────────────

#[test]
fn test_reactivate_from_active_is_error() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(1);
    let actor = store.register(pk, "Alice", ActorType::Human).unwrap();

    let err = store.reactivate(actor.id(), "reason").unwrap_err();
    match err {
        EventGraphError::InvalidTransition { .. } => {}
        other => panic!("expected InvalidTransition, got {:?}", other),
    }
}

#[test]
fn test_reactivate_not_found() {
    let mut store = InMemoryActorStore::new();
    let id = eventgraph::types::ActorId::new("actor_nonexistent").unwrap();
    let err = store.reactivate(&id, "reason").unwrap_err();
    match err {
        EventGraphError::ActorNotFound { .. } => {}
        other => panic!("expected ActorNotFound, got {:?}", other),
    }
}

// ── List ──────────────────────────────────────────────────────────────

#[test]
fn test_list() {
    let mut store = InMemoryActorStore::new();
    for i in 1u8..=5 {
        store.register(test_public_key(i), "Actor", ActorType::Human).unwrap();
    }

    let page = store.list(&ActorFilter {
        status: None,
        actor_type: None,
        limit: 10,
        after: None,
    }).unwrap();
    assert_eq!(page.items.len(), 5);
}

#[test]
fn test_list_with_status_filter() {
    let mut store = InMemoryActorStore::new();
    store.register(test_public_key(1), "Active1", ActorType::Human).unwrap();
    store.register(test_public_key(2), "Active2", ActorType::Human).unwrap();
    let a3 = store.register(test_public_key(3), "ToBeSuspended", ActorType::Human).unwrap();

    store.suspend(a3.id(), "reason").unwrap();

    let page = store.list(&ActorFilter {
        status: Some(ActorStatus::Active),
        actor_type: None,
        limit: 10,
        after: None,
    }).unwrap();
    assert_eq!(page.items.len(), 2);
}

#[test]
fn test_list_with_type_filter() {
    let mut store = InMemoryActorStore::new();
    store.register(test_public_key(1), "Human", ActorType::Human).unwrap();
    store.register(test_public_key(2), "AI", ActorType::AI).unwrap();
    store.register(test_public_key(3), "System", ActorType::System).unwrap();

    let page = store.list(&ActorFilter {
        status: None,
        actor_type: Some(ActorType::AI),
        limit: 10,
        after: None,
    }).unwrap();
    assert_eq!(page.items.len(), 1);
}

#[test]
fn test_list_pagination() {
    let mut store = InMemoryActorStore::new();
    for i in 1u8..=5 {
        store.register(test_public_key(i), "Actor", ActorType::Human).unwrap();
    }

    // Page 1
    let page1 = store.list(&ActorFilter {
        status: None,
        actor_type: None,
        limit: 2,
        after: None,
    }).unwrap();
    assert_eq!(page1.items.len(), 2);
    assert!(page1.has_more);

    // Page 2
    let page2 = store.list(&ActorFilter {
        status: None,
        actor_type: None,
        limit: 2,
        after: page1.cursor,
    }).unwrap();
    assert_eq!(page2.items.len(), 2);
}

// ── Actor getters ─────────────────────────────────────────────────────

#[test]
fn test_actor_getters() {
    let mut store = InMemoryActorStore::new();
    let pk = test_public_key(42);
    let actor = store.register(pk, "Alice", ActorType::AI).unwrap();

    assert!(!actor.id().value().is_empty());
    let _ = actor.public_key();
    assert!(actor.created_at_nanos() > 0);
    let _ = actor.metadata();
}
